sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py

```diff
@@ -0,0 +1,758 @@
+        if self.do_early_stopping_:
+            subsample_size = 10000  # should we expose this parameter?
+            indices = np.arange(X_binned_train.shape[0])
+            if X_binned_train.shape[0] > subsample_size:
+                indices = rng.choice(indices, subsample_size)
+            X_binned_small_train = X_binned_train[indices]
+            y_small_train = y_train[indices]
+            # Predicting is faster on C-contiguous arrays.
+            X_binned_small_train = np.ascontiguousarray(X_binned_small_train)
+
+        if self.verbose:
+            print("Fitting gradient boosted rounds:")
+
+        # initialize raw_predictions: those are the accumulated values
+        # predicted by the trees for the training data. raw_predictions has
+        # shape (n_trees_per_iteration, n_samples) where
+        # n_trees_per_iterations is n_classes in multiclass classification,
+        # else 1.
+        n_samples = X_binned_train.shape[0]
+        self._baseline_prediction = self.loss_.get_baseline_prediction(
+            y_train, self._n_trees_per_iteration)
+        raw_predictions = np.zeros(
+            shape=(self._n_trees_per_iteration, n_samples),
+            dtype=self._baseline_prediction.dtype
+        )
+        raw_predictions += self._baseline_prediction
+
+        # initialize gradients and hessians (empty arrays).
+        # shape = (n_trees_per_iteration, n_samples).
+        gradients, hessians = self.loss_.init_gradients_and_hessians(
+            n_samples=n_samples,
+            prediction_dim=self._n_trees_per_iteration
+        )
+
+        # predictors is a matrix (list of lists) of TreePredictor objects
+        # with shape (n_iter_, n_trees_per_iteration)
+        self._predictors = predictors = []
+
+        # scorer_ is a callable with signature (est, X, y) and calls
+        # est.predict() or est.predict_proba() depending on its nature.
+        if self.scoring != 'loss':
+            self.scorer_ = check_scoring(self, self.scoring)
+        else:
+            self.scorer_ = None
+        self.train_score_ = []
+        self.validation_score_ = []
+        if self.do_early_stopping_:
+            # populate train_score and validation_score with the predictions
+            # of the initial model (before the first tree)
+            self._check_early_stopping(X_binned_small_train, y_small_train,
+                                       X_binned_val, y_val)
+
+        for iteration in range(self.max_iter):
+
+            if self.verbose:
+                iteration_start_time = time()
+                print("[{}/{}] ".format(iteration + 1, self.max_iter),
+                      end='', flush=True)
+
+            # Update gradients and hessians, inplace
+            self.loss_.update_gradients_and_hessians(gradients, hessians,
+                                                     y_train, raw_predictions)
+
+            predictors.append([])
```

Comment:
It looks surprising at first. Maybe a small comment to explain why there is an empty list appended.